File: android/app/src/main/kotlin/ai/gowtam/oak/features/chat/CandidatesTsv.kt
package ai.gowtam.oak.features.chat

import ai.gowtam.oak.wire.BaseStats
import ai.gowtam.oak.wire.CandidateRow
import ai.gowtam.oak.wire.JsonScalar

/**
 * Visible candidate rows → tab-separated values (TBL-US-4).
 *
 * The caller passes the already-visible set (after sort / filter /
 * in-table pin). Hidden remainder is never included (TBL-BR-1).
 * Empty input → empty string (TBL-AC-4.4).
 */

private val STAT_ORDER = listOf("hp", "attack", "defense", "special_attack", "special_defense", "speed")

/** Spreadsheet-paste labels — match the candidate table (Atk/Def/Spe). */
private val STAT_LABELS = mapOf(
    "hp" to "HP",
    "attack" to "Atk",
    "defense" to "Def",
    "special_attack" to "SpA",
    "special_defense" to "SpD",
    "speed" to "Spe",
)

/** Spreadsheet paste of the currently visible candidate table. */
fun candidatesToTsv(visibleRows: List<CandidateRow>): String {
    if (visibleRows.isEmpty()) return ""

    val hasBase = visibleRows.any { it.baseStats != null }
    val keyStatKeys = collectKeyStatKeys(visibleRows)
    val hasKeyStats = !hasBase && keyStatKeys.isNotEmpty()
    val hasAbility = visibleRows.any { !it.ability.isNullOrEmpty() }

    val headers = mutableListOf("Name", "Types")
    if (hasBase) {
        headers += STAT_ORDER.map { STAT_LABELS.getValue(it) }
    } else if (hasKeyStats) {
        headers += keyStatKeys.map { statHeader(it) }
    }
    if (hasAbility) headers += "Ability"

    val body = visibleRows.map { row ->
        val cells = mutableListOf(row.name, row.types.joinToString("/"))
        if (hasBase) {
            val stats = row.baseStats
            for (key in STAT_ORDER) {
                cells += if (stats != null) baseStatValue(stats, key) else ""
            }
        } else if (hasKeyStats) {
            for (key in keyStatKeys) {
                val value = row.keyStats?.get(key)
                cells += if (value == null) "" else scalarText(value)
            }
        }
        if (hasAbility) cells += row.ability.orEmpty()
        cells.joinToString("\t")
    }

    return (listOf(headers.joinToString("\t")) + body).joinToString("\n")
}

private fun collectKeyStatKeys(rows: List<CandidateRow>): List<String> {
    val keys = mutableListOf<String>()
    val seen = mutableSetOf<String>()
    for (row in rows) {
        val stats = row.keyStats ?: continue
        for (key in stats.keys) {
            if (seen.add(key)) keys += key
        }
    }
    return keys
}

private fun statHeader(key: String): String {
    if (key == "hp") return "HP"
    return key.split("-").joinToString(" ") { word ->
        if (word.isEmpty()) word else word[0].uppercaseChar() + word.substring(1)
    }
}

private fun baseStatValue(stats: BaseStats, key: String): String = when (key) {
    "hp" -> stats.hp.toString()
    "attack" -> stats.atk.toString()
    "defense" -> stats.def.toString()
    "special_attack" -> stats.spa.toString()
    "special_defense" -> stats.spd.toString()
    "speed" -> stats.spe.toString()
    else -> ""
}

private fun scalarText(value: JsonScalar): String = when (value) {
    is JsonScalar.Str -> value.v
    is JsonScalar.IntVal -> value.v.toString()
    is JsonScalar.DoubleVal -> value.v.toString()
    is JsonScalar.BoolVal -> value.v.toString()
    JsonScalar.Null -> ""
}

File: web/src/lib/candidates-tsv.ts
/**
 * Visible candidate rows → tab-separated values (TBL-US-4).
 *
 * The caller passes the already-visible set (after sort / filter /
 * in-table pin). Hidden remainder is never included (TBL-BR-1).
 * Empty input → empty string (TBL-AC-4.4).
 */

import type { Candidates } from "@/agent/schemas";

export type CandidateRow = Candidates["shown"][number];

const STAT_ORDER = [
  "hp",
  "attack",
  "defense",
  "special_attack",
  "special_defense",
  "speed",
] as const;

const STAT_LABELS: Record<(typeof STAT_ORDER)[number], string> = {
  hp: "HP",
  attack: "Attack",
  defense: "Defense",
  special_attack: "SpA",
  special_defense: "SpD",
  speed: "Speed",
};

function humanize(slug: string): string {
  return slug
    .split("-")
    .map((w) => (w.length === 0 ? w : w[0]!.toUpperCase() + w.slice(1)))
    .join(" ");
}

function statHeader(key: string): string {
  if (key === "hp") return "HP";
  return humanize(key);
}

function collectKeyStatKeys(rows: CandidateRow[]): string[] {
  const keys: string[] = [];
  const seen = new Set<string>();
  for (const row of rows) {
    if (!row.key_stats) continue;
    for (const key of Object.keys(row.key_stats)) {
      if (seen.has(key)) continue;
      seen.add(key);
      keys.push(key);
    }
  }
  return keys;
}

/** Spreadsheet paste of the currently visible candidate table. */
export function candidatesToTsv(rows: CandidateRow[]): string {
  if (rows.length === 0) return "";

  const hasBase = rows.some((row) => row.base_stats != null);
  const keyStatKeys = collectKeyStatKeys(rows);
  const hasKeyStats = !hasBase && keyStatKeys.length > 0;
  const hasAbility = rows.some((row) => row.ability != null && row.ability !== "");

  const headers = ["Name", "Types"];
  if (hasBase) {
    headers.push(...STAT_ORDER.map((k) => STAT_LABELS[k]));
  } else if (hasKeyStats) {
    headers.push(...keyStatKeys.map(statHeader));
  }
  if (hasAbility) headers.push("Ability");

  const body = rows.map((row) => {
    const cells = [row.name, row.types.join("/")];
    if (hasBase) {
      for (const k of STAT_ORDER) {
        cells.push(row.base_stats != null ? String(row.base_stats[k]) : "");
      }
    } else if (hasKeyStats) {
      for (const key of keyStatKeys) {
        const value = row.key_stats?.[key];
        cells.push(value == null ? "" : String(value));
      }
    }
    if (hasAbility) cells.push(row.ability ?? "");
    return cells.join("\t");
  });

  return [headers.join("\t"), ...body].join("\n");
}
